lib/SILOptimizer/IPO/DeadFunctionElimination.cpp

```diff
@@ -365,10 +365,12 @@ class DeadFunctionElimination : FunctionLivenessComputation {
   /// Removes all dead methods from vtables and witness tables.
   void removeDeadEntriesFromTables() {
     for (SILVTable &vTable : Module->getVTableList()) {
-      vTable.removeEntries_if([this](SILVTable::Pair &entry) -> bool {
+      vTable.removeEntries_if([&](SILVTable::Pair &entry) -> bool {
         if (!isAlive(entry.second)) {
           DEBUG(llvm::dbgs() << "  erase dead vtable method " <<
                 entry.second->getName() << "\n");
+          // Update cache:
+          Module->removeFromVTableCache(&vTable, entry.first);
```

Comment:
I still think it would be cleaner, because all the cache handling is done inside SILVTable.
You can get the module from the SILFunction of the vtable entry (which you are about to delete).
To solve the forward-declaration problem, you can define a private SILVtable helper function, which is defined in the cpp file.